src/environmentd/src/bin/environmentd/main.rs

```diff
@@ -742,7 +775,42 @@ fn run(mut args: Args) -> Result<(), anyhow::Error> {
                     }))
                     .context("creating kubernetes orchestrator")?,
             );
-            let secrets_controller = Arc::clone(&orchestrator);
+            let secrets_controller: Arc<dyn SecretsController> =
+                match args.secrets_controller {
+                    SecretsControllerKind::Kubernetes => {
+                        let sc = Arc::clone(&orchestrator);
+                        let sc: Arc<dyn SecretsController> = sc;
+                        sc
+                    }
+                    SecretsControllerKind::AwsSecretsManager => {
+                        // One-time migrate any existing kubernetes secrets
+                        let kubernetes_secrets_controller = &orchestrator;
+                        let kubernetes_secrets: BTreeSet<GlobalId> = runtime.block_on(kubernetes_secrets_controller.list())?.into_iter().collect();
+
+                        let aws_secrets_controller = Arc::new(
+                            runtime.block_on(AwsSecretsController::new(
+                                args.environment_id.clone(),
+                                args.aws_secrets_controller_tags
+                                    .into_iter()
+                                    .map(|tag| (tag.key, tag.value))
+                                    .collect(),
+                            )),
+                        );
+                        let aws_secrets: BTreeSet<GlobalId> = runtime.block_on(aws_secrets_controller.list())?.into_iter().collect();
+
+                        let kubernetes_secrets_reader = kubernetes_secrets_controller.reader();
+                        let secrets_to_migrate = kubernetes_secrets.difference(&aws_secrets);
+                        for secret in secrets_to_migrate {
+                            let data = runtime.block_on(kubernetes_secrets_reader.read(secret.to_owned()))?;
+                            runtime.block_on(aws_secrets_controller.ensure(secret.to_owned(), &data))?;
+                        }
+                        for secret in kubernetes_secrets {
+                            runtime.block_on(kubernetes_secrets_controller.delete(secret))?;
```

Comment:
I have this as a separate loop rather than doing it immediately after each secret just in case there is a failure after partially migrating, but before deleting the K8S secrets. This might not be a realistic scenario, though, since it would only be useful if at least one secret has been written successfully to AWS, but not all of them.